src/neurostatslib/base_class.py

```diff
@@ -0,0 +1,578 @@
+    - [`GLMRecurrent`](../glm/#neurostatslib.glm.GLMRecurrent): A recurrent GLM implementation.
+    """
+
+    FLOAT_EPS = jnp.finfo(jnp.float32).eps
+
+    @abc.abstractmethod
+    def fit(self, X: Union[NDArray, jnp.ndarray], y: Union[NDArray, jnp.ndarray]):
+        """Fit the model to neural activity."""
+        pass
+
+    @abc.abstractmethod
+    def predict(self, X: Union[NDArray, jnp.ndarray]) -> jnp.ndarray:
+        """Predict rates based on fit parameters."""
+        pass
+
+    @abc.abstractmethod
+    def score(
+        self,
+        X: Union[NDArray, jnp.ndarray],
+        y: Union[NDArray, jnp.ndarray],
+        # may include score_type or other additional model dependent kwargs
+        **kwargs,
+    ) -> jnp.ndarray:
+        """Score the predicted firing rates (based on fit) to the target neural activity."""
+        pass
+
+    @abc.abstractmethod
+    def simulate(
+        self,
+        random_key: jax.random.PRNGKeyArray,
+        feed_forward_input: Union[NDArray, jnp.ndarray],
+    ):
+        """Simulate neural activity in response to a feed-forward input and recurrent activity."""
+        pass
+
+    @staticmethod
+    def _convert_to_jnp_ndarray(
+        *args: Union[NDArray, jnp.ndarray], data_type: Optional[jnp.dtype] = None
+    ) -> Tuple[jnp.ndarray, ...]:
+        """Convert provided arrays to jnp.ndarray of specified type.
+
+        Parameters
+        ----------
+        *args :
+            Input arrays to convert.
+        data_type :
+            Data type to convert to. Default is None, which means that the data-type
+            is inferred from the input.
+
+        Returns
+        -------
+        :
+            Converted arrays.
+        """
+        return tuple(jnp.asarray(arg, dtype=data_type) for arg in args)
+
+    @staticmethod
+    def _has_invalid_entry(array: jnp.ndarray) -> bool:
+        """Check if the array has nans or infs.
+
+        Parameters
+        ----------
+        array:
+            The array to be checked.
+
+        Returns
+        -------
+            True if a nan or an inf is present, False otherwise
+
+        """
+        return (jnp.isinf(array) | jnp.isnan(array)).any()
+
+    @staticmethod
+    def _check_and_convert_params(
+        params: Tuple[ArrayLike, ArrayLike], data_type: Optional[jnp.dtype] = None
+    ) -> Tuple[jnp.ndarray, jnp.ndarray]:
+        """
+        Validate the dimensions and consistency of parameters and data.
+
+        This function checks the consistency of shapes and dimensions for model
+        parameters.
+        It ensures that the parameters and data are compatible for the model.
+
+        """
+        if not is_sequence(params):
+            raise TypeError("Initial parameters must be array-like!")
+
+        if len(params) != 2:
+            raise ValueError("Params needs to be array-like of length two.")
+
+        try:
+            params = jnp.asarray(params[0], dtype=data_type), jnp.asarray(
```

Comment:
that's a great point, I should move some of the glm specific checks in the  GLM class, so that we keep supporting the pytrees;  the conversion to jax utils should also call tree_utils probably, so that it converts any tree-like structure

Reply:
I vote for holding off for now, since we will work on the pytree version of the code